htlcswitch/mock.go

```diff
@@ -699,82 +701,83 @@ func (f *mockChannelLink) UpdateShortChanID() (lnwire.ShortChannelID, error) {
 
 var _ ChannelLink = (*mockChannelLink)(nil)
 
-type mockInvoiceRegistry struct {
-	sync.Mutex
-
-	invoices   map[lntypes.Hash]channeldb.Invoice
-	finalDelta uint32
-}
-
-func newMockRegistry(minDelta uint32) *mockInvoiceRegistry {
-	return &mockInvoiceRegistry{
-		finalDelta: minDelta,
-		invoices:   make(map[lntypes.Hash]channeldb.Invoice),
+func newDB() (*channeldb.DB, func(), error) {
+	// First, create a temporary directory to be used for the duration of
+	// this test.
+	tempDirName, err := ioutil.TempDir("", "channeldb")
+	if err != nil {
+		return nil, nil, err
 	}
-}
 
-func (i *mockInvoiceRegistry) LookupInvoice(rHash lntypes.Hash) (channeldb.Invoice, uint32, error) {
-	i.Lock()
-	defer i.Unlock()
+	// Next, create channeldb for the first time.
+	cdb, err := channeldb.Open(tempDirName)
+	if err != nil {
+		os.RemoveAll(tempDirName)
+		return nil, nil, err
+	}
 
-	invoice, ok := i.invoices[rHash]
-	if !ok {
-		return channeldb.Invoice{}, 0, fmt.Errorf("can't find mock "+
-			"invoice: %x", rHash[:])
+	cleanUp := func() {
+		cdb.Close()
+		os.RemoveAll(tempDirName)
 	}
 
-	return invoice, i.finalDelta, nil
+	return cdb, cleanUp, nil
 }
 
-func (i *mockInvoiceRegistry) SettleInvoice(rhash lntypes.Hash,
-	amt lnwire.MilliSatoshi) error {
+type mockInvoiceRegistry struct {
+	settleChan chan lntypes.Hash
 
-	i.Lock()
-	defer i.Unlock()
+	registry *invoices.InvoiceRegistry
 
-	invoice, ok := i.invoices[rhash]
-	if !ok {
-		return fmt.Errorf("can't find mock invoice: %x", rhash[:])
+	cleanup func()
+}
+
+func newMockRegistry(minDelta uint32) *mockInvoiceRegistry {
+	cdb, cleanup, err := newDB()
+	if err != nil {
+		panic(err)
 	}
 
-	if invoice.Terms.State == channeldb.ContractSettled {
-		return nil
+	decodeExpiry := func(invoice string) (uint32, error) {
+		return 3, nil
 	}
 
-	invoice.Terms.State = channeldb.ContractSettled
-	invoice.AmtPaid = amt
-	i.invoices[rhash] = invoice
+	registry := invoices.NewRegistry(cdb, decodeExpiry)
```

Comment:
The reason that a real instance is used here, is to be able to test the behavior of the link in combination with the actual registry implementation. This could be mocked out, but the mock would need to be implemented and controlled carefully to make sure it does the same thing. I think using the actual implementation is a safer solution. If performance is a concern, I'd prefer mocking the db layer underneath the registry.

Reply:
Understood, thanks for the explanation.